### PR TITLE
Cleanup image Makefiles and add security check.

### PR DESCRIPTION
**Pull request by coryrc**
Should upload images once they've been reviewed and approved. Make it
a little harder to not do that.

Clean and simplify the Makefiles. Add convenience function for
launching a shell in your local image.

### CODE REVIEW

File: images/README.md [hunk 1]
```diff
@@ -1,3 +1,20 @@
 # Prow Job Images
 
 This directory contains custom Docker images used by our Prow jobs.
+
+## Building and publishing a new image
+
+To build and push a new image, you can run `make push` from an image directory. This should only be
+done once a PR is approved and from the master branch.
```
Comment (mattmoor) [inline images/README.md:8]: Format markdown:
```suggestion
To build and push a new image, you can run `make push` from an image directory.
This should only be done once a PR is approved and from the master branch.
```

File: images/README.md [hunk 2]
```diff
@@ -1,3 +1,20 @@
...
+
+Users typically keep their working directory at the repo root and run relative make commands, like
+`make -C images/prow-tests push`, but for brevity's sake this document just mentions the short way.
```
Comment (mattmoor) [inline images/README.md:11]: Format markdown:
```suggestion
Users typically keep their working directory at the repo root and run relative
make commands, like `make -C images/prow-tests push`, but for brevity's sake
this document just mentions the short way.
```

File: images/README.md [hunk 3]
```diff
@@ -1,3 +1,20 @@
...
+
+For testing purposes, you can build with --no-cache, which does not rerun a step in the Dockerfile
+if the text in the Dockerfile does not change (which, note, if you are doing `go install` or
+similar, and you update files, the image will not change! Either `make build` to rebuild everything
+or add/remove a no-op flag to the command you care about). Do this by running `make
+iterative-build`. Run `make iterative-shell` to get a shell in this image.
```
Comment (mattmoor) [inline images/README.md:17]: Format markdown:
```suggestion
For testing purposes, you can build with --no-cache, which does not rerun a step
in the Dockerfile if the text in the Dockerfile does not change (which, note, if
you are doing `go install` or similar, and you update files, the image will not
change! Either `make build` to rebuild everything or add/remove a no-op flag to
the command you care about). Do this by running `make iterative-build`. Run
`make iterative-shell` to get a shell in this image.
```

File: images/README.md [hunk 4]
```diff
@@ -1,3 +1,20 @@
...
+
+Note that you must have proper permission in the `knative-tests` project to push new images to the
+GCR.
```
Comment (mattmoor) [inline images/README.md:20]: Format markdown:
```suggestion
Note that you must have proper permission in the `knative-tests` project to push
new images to the GCR.
```

File: images/prow-tests/README.md [hunk 5]
```diff
@@ -1,26 +1,15 @@
 # Prow Test Job Image
 
-This directory contains the custom Docker image used by our Prow test jobs. A
-fork of this image with go1.12 installed is defined in
-[`prow-tests-go112`](../prow-tests-go112)
+This directory contains the custom Docker image used by our Prow test jobs.
+Some forks of this images exist under images/.
```
Comment (mattmoor) [inline images/prow-tests/README.md:4]: Format markdown:
```suggestion
This directory contains the custom Docker image used by our Prow test jobs. Some
forks of this images exist under images/.
```

File: common.mk [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Do not allow server update from wrong branch or dirty working space
+# In emergency, could easily edit this file, deleting all these lines
+confirm-master:
+	@if git diff-index --quiet HEAD; then true; else echo "Git working space is dirty -- will not update server"; false; fi;
+# TODO(chizhg): change to `git branch --show-current` after we update the Git version in prow-tests image.
```
Comment (chaodaiG) [inline common.mk:19]: @chizhg , does this PR affect your TODO?

**Comment by coryrc** [comment]
@googlebot rescan

File: common.mk (hunk 6 above)
Comment (chizhg) [reply common.mk:19]: No, the TODO should still be kept as the Git version in the prow-tests image is still pretty old.

**Comment by chizhg** [comment]
Need to get https://github.com/knative/test-infra/pull/1989 in first..

**Pull request merged**